File: solutions/ruby/meetup/1/meetup.rb
require 'date'

class Meetup
    def initialize(month, year)
        @month = month
        @year = year
    end

    def day(dayOfWeek, type)
        firstDay = Date.new(@year, @month, 1)
        lastDay = Date.new(@year, @month, -1)
        case type
        when :first
            firstDay.upto(lastDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s
                    return date
                end
            end
        when :second
            firstDay.upto(lastDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s
                    return date + 7
                end
            end
        when :third
            firstDay.upto(lastDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s
                    return date + 14
                end
            end
        when :fourth
            firstDay.upto(lastDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s
                    return date + 21
                end
            end
        when :teenth
            firstDay.upto(lastDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s && date.day >= 13 && date.day <= 19
                    return date
                end
            end
        when :last
            lastDay.downto(firstDay) do |date|
                if date.strftime("%A").downcase == dayOfWeek.to_s
                    return date
                end
            end
        end
    end
end

File: solutions/elixir/meetup/1/lib/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @weekdays [:monday, :tuesday, :wednesday, :thursday, :friday, :saturday, :sunday] |> Enum.with_index(1) |> Enum.into(%{})

  @schedule_range %{
    first: 1..7,
    second: 8..14,
    third: 15..21,
    fourth: 22..28,
    teenth: 13..19,
  }
  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date()
  def meetup(year, month, weekday, schedule) do
    weekday = @weekdays[weekday]
    Map.get(@schedule_range, schedule, last_seven(year, month))
    |> Enum.map(&(Date.from_erl!({year, month, &1})))
    |> Enum.filter(&(Date.day_of_week(&1) == weekday))
    |> List.first
  end

  defp last_seven(year, month) do
    last_day = Date.from_erl!({year, month, 1}) |> Date.end_of_month
    first_day = last_day.day - 6
    first_day..last_day.day
  end
end
